lib/player.js

```diff
@@ -5524,6 +5588,49 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
     }
   }
 
+  /**
+   * Callback for video progress events
+   *
+   * @private
+   */
+  onVideoProgress_() {
+    if (!this.video_) {
+      return;
+    }
+    let hasNewCompletionPercent = false;
+    const completionRatio = this.video_.currentTime / this.video_.duration;
+    if (!isNaN(completionRatio)) {
+      const percent = Math.round(100 * completionRatio);
+      if (isNaN(this.completionPercent_)) {
+        this.completionPercent_ = percent;
+        hasNewCompletionPercent = true;
+      } else {
+        const newCompletionPercent = Math.max(this.completionPercent_, percent);
+        if (this.completionPercent_ != newCompletionPercent) {
+          this.completionPercent_ = newCompletionPercent;
+          hasNewCompletionPercent = true;
+        }
+      }
+    }
+    if (hasNewCompletionPercent) {
+      let event;
+      if (this.completionPercent_ == 0) {
```

Comment:
Hm.
According to MDN, at the slowest, we can expect around 4 `timeupdate` events per second.
For a very short clip (<25 seconds), if the events are happening at the slowest expected rate, there is the possibility that the completion percent might go up by more than 1 between two events, which could cause us to skip over a quartile event...
I think this would be most likely to happen with the started event, since all that would really need to happen would be for the first `timeupdate` event to happen at 0.51% progress (so it gets rounded up to 1%).

I don't know how realistic that combination of factors would be in practice, though.